### PR TITLE
Adding TRUST4 module

### LINKED ISSUE #4307: new module: trust4
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by Joaodemeirelles**
Draft for the TRUST4 module addition.

Since TRUST4 can be execute with single-end reads, paired-end reads and a bam file, I chose to handle this by testing if the channel exists (to differ between bam and reads), and the meta.single_end key to identify paired or single-end.

## PR checklist

Closes #4307 
- [X] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by mapo9** [comment]
Hi Joao @Joaodemeirelles,

thanks for your work in adding this module!
I am also very interested to have this and would have some time now to work on it.

Can you tell me what still needs to be done?

Also, did you think about implementing tests yet?

**Comment by mapo9** [comment]
Hi, I took this PR up again as I want to make use of the tool.

I experience some problems that I am unsure on how to deal with them. 
First, nf-test fails as some files stay empty in the small test dataset. As linting failed due to md5sums for empty files, I removed the md5sums for these files. This made nf-test fail as the snapshots are now different due to the missing md5sums. 

Also, singularity fails because there is no singularity container for the newest version of the tool and some output seems to have changed so the md5sums differ as well from the docker & conda run with the newest version. 

Does anyone has an idea how to solve this?

**Comment by mapo9** [comment]
I was able to solve the problems above and think that the PR is now ready for review :)

**Comment by Joaodemeirelles** [comment]
> I was able to solve the problems above and think that the PR is now ready for review :)

Great! Sorry I could not help with the tests, thank you so much for the changes!

File: modules/nf-core/trust4/main.nf [hunk 1]
```diff
@@ -0,0 +1,73 @@
+process TRUST4 {
+    tag "$meta.id"
+    label 'process_medium'
+
+    conda "bioconda::trust4=1.0.13"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'biocontainers/trust4:1.0.13--h43eeafb_0':
+        'biocontainers/trust4:1.0.13--h43eeafb_0' }"
```
Comment (FriederikeHanssen) [inline modules/nf-core/trust4/main.nf:8]:
```suggestion
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/trust4:1.0.13--h43eeafb_0':
        'biocontainers/trust4:1.0.13--h43eeafb_0' }"
```

File: modules/nf-core/trust4/main.nf [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    input:
+    tuple val(meta), path(bam), path(reads)
+    path(fasta)
+    path(ref)
```
Comment (FriederikeHanssen) [inline modules/nf-core/trust4/main.nf:13]: we need meta1 and meta2 here. What is the difference between fasta and ref, how about: vdj_reference?

Comment (mapo9) [reply modules/nf-core/trust4/main.nf:13]: ref is an additional reference required when you dont start from bam but fastq
fasta is what they call their reference basically.
I wanted to stay in line with their naming here, but am also fine with renaming them.

**Review by ggabernet — APPROVED** [review]
Looks good to me now! There's only the `trinity` pytest that I'm not sure they should be added here

**Pull request merged**